File: LeetCode 0314 Binary Tree Vertical Order Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object): # 100%, 68%
    def verticalOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        # 定义一个性质：index，左子节点的index是其根的index-1，右则+1。index相同则属同一列，同列按level排序，所以在外进行一个普通层序遍历
        if not root: return []
        queue = [root]
        columns = {} # index -> [node_values]
        indexes = {} # node -> node.index
        indexes[root] = 0
        while queue:
            level_size = len(queue)
            for _ in range(level_size):
                node = queue.pop(0)
                index = indexes[node]
                # put(index, node.val)
                if index not in columns: columns[index] = [node.val]
                else: columns[index].append(node.val)
                # push children and record their indexes
                if node.left: 
                    queue.append(node.left)
                    indexes[node.left] = index - 1
                if node.right:
                    queue.append(node.right)
                    indexes[node.right] = index + 1
        return [v for k, v in sorted(columns.items())]

File: LeetCode 0314 Binary Tree Vertical Order Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> verticalOrder(TreeNode root) { // 94, 78
        if (root == null)
            return new LinkedList<List<Integer>>();
        
        Queue<TreeNode> q = new LinkedList<TreeNode>(); // 层序遍历需要的queue
        List<List<Integer>> ans = new LinkedList<List<Integer>>(); // 运算结果
        Map<TreeNode, Integer> nodeOffsets = new HashMap<TreeNode, Integer>(); // Node对应的offset值
        Map<Integer, List<Integer>> offsetVal = new HashMap<Integer, List<Integer>>(); // offset值对应的所有node值，按entry转换成答案
        int minKey = 0;
        
        nodeOffsets.put(root, 0);
        q.offer(root);
        while (!q.isEmpty()) {
            TreeNode node = q.poll();
            int nodeOffset = nodeOffsets.get(node);
            
            if (!offsetVal.containsKey(nodeOffset))
                offsetVal.put(nodeOffset, new LinkedList<Integer>());
            offsetVal.get(nodeOffset).add(node.val);
            
            if (node.left != null) {
                q.offer(node.left);
                nodeOffsets.put(node.left, nodeOffset - 1);
                if (nodeOffset - 1 < minKey)
                    minKey = nodeOffset - 1;
            }
            if (node.right != null) {
                q.offer(node.right);
                nodeOffsets.put(node.right, nodeOffset + 1);
            }
        }
        
        while (offsetVal.containsKey(minKey)) {
            ans.add(offsetVal.get(minKey));
            minKey++;
        }
        
        return ans;
    }
}